cassiopeia/data/CassiopeiaTree.py

```diff
@@ -1607,12 +1613,19 @@ def collapse_mutationless_edges(
                 of the tree
 
         Raises:
-            CassiopeiaTreeError if the tree has not been initialized.
+            CassiopeiaTreeError if the tree has not been initialized or if
+            a node does not have character states
         """
         if infer_ancestral_characters:
             self.reconstruct_ancestral_characters()
 
         for n in list(self.depth_first_traverse_nodes(postorder=True)):
+            if len(self.get_character_states(n)) == 0:
+                raise CassiopeiaTreeError(
+                    f"Character states empty at node {n}, annotate"
+                    " character states or infer ancestral characters before"
+                    " mutationless edges can be collaped."
```

Comment:
Two things:

(1) First, let's be more clear that you can set `infer_ancestral_states=True` in this method to take care of this issue.
(2) Second, I'm still not sure why the `character_states` attribute would ever be empty at a node since `populate_tree` always gets run.

Reply:
So the issue here is that if the user erroneously believes that the internal nodes are annotated when they are indeed not, then if they run `collapse_mutationless_edges` then they'll end up collapsing the entire tree which would definitely mess up their pipelines if they aren't aware of it, so we just spit an error to prevent headache and debugging on their part. What's meant by "character states are empty" here is that the character states are the empty list, not necessarily that the attribute doesn't exist. I'll make it more explicit that they can set the argument "infer_ancestral_states=True" to fix this. Thanks!

Reply:
This case is tested in the second case of `test_states_no_annotation`